### PR TITLE
feat(`V6`): Group Rewrite + Nested Selection

### LINKED ISSUE #3000: Upvote: Layers
At the Wiki there is "_Support for multiple layers (for performance)"_ on the Roadmap.

I would like to upvote this feature. 

It is not only helpfull for performance reasons, but can be very useful for any  kind of drawing application. Basically it offers a powerull way to organize the drawings, steers which objects are currently visible/selectable/changeable.

For those, who are not familiar with the concept of layers, here is a short introduction: 
https://www.youtube.com/watch?v=CuKNa4KBC8s
This video shows some powerful examples how layers can be used:
https://www.youtube.com/watch?v=tEzBAVPQUF0
(_Edit: I updated the video-links to better ones. This was the original link_: https://www.youtube.com/watch?v=mE82knjW7IA)

I thought a bit about the functionality of a layer feature in fabrics. If you are interested, we can discuss it here.

PS: I appreciate the faster release policy very much.

### LINKED ISSUE #3012: Do more actions with items in groups
This is proposal for a possible feature, interaction with sub objects of grouped objects.

Actually fabricjs can just detects click over them.
We need to define:
1) actions available with grouped objects
2) method of activating this objects
3) having both internal and normalized corner coords
4) .....

Discuss Pro and Cons of keeping this in the library or keeping at application level

I throw some example:
moving object inside a group without ungrouping.
problems: relative coordinates and updating group dimensions.

Nested groups, full support of basic functionalities we have now

That is a long term feature.

@kangax  @kienz @inssein @samuelhorwitz

### LINKED ISSUE #6499: Bounding box for group doesn't update when grouped objects are changed
I'm having a problem with the bounding box of a group, when objects inside the group are modified. When I change the stroke width and width/height of grouped objects, the bounding box does not automatically update.

[Here's the question on Stack Overflow with a working snippet.](https://stackoverflow.com/questions/63256748/fabric-js-3-6-bounding-box-for-group-doesnt-update-when-grouped-objects-are-c)

[The same example is also posted at Codepen.](https://codepen.io/green_footballs/pen/GRZRZJa?editors=0011)

### LINKED ISSUE #7130: Object geometry methods for grouped objects
## Version
latest

## Test Case
http://jsfiddle.net/mkdve1x9/15/

## Information about environment
Browser.

## Steps to reproduce
1. Create a `Group`/`ActiveSelection`
2. Try using a method from [`object_geomerty`](https://github.com/fabricjs/fabric.js/blob/2eabc92a3221dd628576b1bb029a5dc1156bdc06/src/mixins/object_geometry.mixin.js#L191) on an object that belongs to the group

## Expected Behavior
Should work for nested objects without additional code

## Actual Behavior
It returns wrong values because the object's transform matrix is used and is relative to the group

### LINKED ISSUE #7142: `absolutePositioned` with objects nested in groups
## Version
latest

## Test Case
http://jsfiddle.net/2185reh6/19/

## Information about environment
browser chrome

## Steps to reproduce
Add an `absolutePositioned` clip path to an object inside a group

## Expected Behavior
Clip path should be positioned relative to canvas, http://fabricjs.com/docs/fabric.Object.html#absolutePositioned

## Actual Behavior
Because group manages it's own plane the clip path is relative to the group

### LINKED ISSUE #7598: Editing text inside groups
## Version
4.6.0

## Test Case

http://jsfiddle.net/pce9h6fv/6/

## Steps to reproduce

1. Open fiddle
2. Click "edit" (calls `text.enterEditing()`)
3. Type something

I am aware that IText editing is not supported when it's in groups https://github.com/fabricjs/fabric.js/issues/3787 and there are custom solutions like here: https://stackoverflow.com/questions/24449481/fabric-js-grouped-itext-not-editable, for example to create a temporary text node https://stackoverflow.com/a/67689690/5125659. 

But this is not very optimal.

I am wondering, if there's a way to route events from a group to particular text node, given that I know, there will be only one such node per group, or it will be more work that fake text node way?

I also attempted to set custom SVG background to IText node, in order to avoid groups at all, but this is not supported either: https://github.com/fabricjs/fabric.js/discussions/7297.

## Expected Behavior

Text is properly shown

## Actual Behavior

Text input is missing, yet the cursor is moving.

Additionally if group is rotated/moved during editing it leaves a cursor trail:

image

### PR DESCRIPTION
**Pull request by ShaMan123**
# Group Rewrite
This PR has been broken down into #7858  #7859 #7860 and will be left open for tracking only.

**This description is updated continuously with relevant information**
You don't need to scroll and read endless comments (unless you are a maintainer :))

## Motivation
#7473 https://github.com/fabricjs/fabric.js/pull/7316 https://github.com/fabricjs/fabric.js/issues/6776 https://github.com/fabricjs/fabric.js/discussions/7136 https://github.com/fabricjs/fabric.js/discussions/7299 https://github.com/fabricjs/fabric.js/issues/7130 https://github.com/fabricjs/fabric.js/issues/7142 #7449 and anything with a `group` tag

closes #6499, closes #3012, closes #3000, closes #7130, closes #7142, closes #7598

Nested selection is a MUST.
Group wasn't easy to use and was very limiting (e.g. doesn't support resizing logic, customize rendering - bg) and confusing (relative coordinate system and transform).

## Current State
Fixed logic across fabric for nested objects to be selectable.
Selection logic needs refactoring to support some cases of nested selection (multiselect), for now use `canvas.setActiveObject`.
Group has been completely rewritten and Layer has been introduced. As part of the rewrite group is now a layout engine, you can use it to customize layout of objects.
Group now respects  `originX` `originY` (still buggy)

### BREAKING
- all awkward methods such as `addWithUpdate` are removed

## Working Examples
- [**CodeSandbox**](https://codesandbox.io/s/jumshw) or [**JSFiddle**](http://jsfiddle.net/Lxjegow3/8/)
- [**Layer**](https://codesandbox.io/s/fabric-app-forked-cxcz3) or in [**JSFiddle**](http://jsfiddle.net/6xjszqtg/1/)

## POC
Tryouts (out of scope, and stale, here to light the way):
- [ICollection 1](https://codesandbox.io/s/yiliv)
- [ICollection 2](https://codesandbox.io/s/fabric-app-forked-cvewn)

## Alpha testing
- cdn: https://cdn.jsdelivr.net/gh/fabricjs/fabric.js@vv6!-build/dist/fabric.js
- npm: `npm i --save fabricjs/fabric.js#v6! && cd ./node_modules/fabric && npm run build`
- yarn pnp: follow the patching instructions in the POC

## To Do

- [ ] refactor selection logic (use `dblClick` to select nested objects? create mechanism for dev to customize selection)
- [ ] refactor to enable n levels of nesting (logic is adequate for 2)  https://github.com/fabricjs/fabric.js/blob/109efe52d1ae6de780389c0dbfefbaf051646a21/src/canvas.class.js#L422 
- [ ] @asturur  rethink `preserveObjectStacking` in favor of a drag image/ghost
- [x] text selection in scaled and rotated group is wrong (`getLocalPointer` 8bf7f92)
- [x] bounding box and controls of object under transformed group (#7758 )
- [ ] phase out coords (can be delayed, now that bbox and controls are fixed)
- [ ] look into active selection - creating from nested objects
- [ ] update tests
- [ ] object caching in some cases leaves traces 
- [ ] refactor object caching
- [ ] caching edge case - nested absolute clip path (see [fiddle](http://jsfiddle.net/prw6g1x5/), try moving group, image should be clipped). Possible fix is to handle a registry of children with abs clip paths **OR** modify the `shouldCache` method.
- [ ] fix group's caching behavior + shadow logic - #7454 
- [ ] support layers under groups?
- [x] ActiveSelection.toGroup use `removeAll` instead
- [x] `getObjectsBoundingBox` - needs work to exactly fit when group is rotated
- [x] initial layout inconsistent with old group
- [ ] initial layout - supporting `originX/Y` 407023cc and `angle`
- [x] `groupSVGElements`
- [x] support fill, stroke
- [ ] from svg - parsing back fill/stroke without adding redundant rect, layout props
- [ ] rethink `_renderObjects` - nested active object will fail with `perPixelTargetFind` because it will not get rendered by group (add logic to `canvas.isTargetTransparent`)

## Look
Video is running in original speed
tiger

## Cutting Edge
The blue circles are rendered by a textbox and because of the rewrite with a few lines of code I got this working
Pay no attention to caching of course

This is **EXTREMELY** exciting because it means that anyone could create custom objects subclassing whatever object they need and have nested selection work out of the box!

ezgif com-gif-maker (13)

### CODE REVIEW

**Comment by ShaMan123** [comment]
@asturur @melchiar this is pretty stable. Still some problems with multiselect that need to be managed and a broader decision on how selection should behave.
Take a look and test it.

**Comment by ShaMan123** [comment]
@asturur regarding bounding box, I saw you talked about moving it to top canvas. I will not work on it until I understand what's the decision. It's a complete mess anyways and deserves a rewrite (especially if we're dropping coords)

**Comment by ShaMan123** [comment]
Say goodbye to the odd creature Group was and **welcome** the new predecessor! Better than ICollection!
6ce776c onwards

**Comment by ShaMan123** [comment]
Layer is now supported

File: build.js [hunk 1]
```diff
@@ -178,6 +178,7 @@ var filesToInclude = [
   'src/shapes/object.class.js',
   'src/mixins/object_origin.mixin.js',
   'src/mixins/object_geometry.mixin.js',
+  'src/mixins/object_ancestry.mixin.js',
```
Comment (ShaMan123) [inline build.js:181]: is this a good name?

File: src/shapes/group.class.js [hunk 2]
```diff
@@ -15,572 +19,733 @@
...
+        }, this);
+        //  clip path as well
+        context.type !== 'initialization' && this.clipPath && !this.clipPath.absolutePositioned
+          && this._adjustObjectPosition(this.clipPath, diff);
+        //  set position
+        this.setPositionByOrigin(newCenter, 'center', 'center');
+        context.type !== 'initialization' && this.callSuper('setCoords');
+        //  fire layout hook and event
+        this.onLayout(context, result);
+        this.fire('layout', {
```
Comment (ShaMan123) [inline src/shapes/group.class.js:438]: should we fire a `modified` event instead?

File: src/static_canvas.class.js [hunk 3]
```diff
@@ -789,7 +789,7 @@
     _onObjectRemoved: function(obj) {
       this.fire('object:removed', { target: obj });
       obj.fire('removed');
-      delete obj.canvas;
+      obj._set('canvas', undefined);
```
Comment (ShaMan123) [inline src/static_canvas.class.js:792]: @asturur please validate this commit is correct

File: src/canvas.class.js [hunk 4]
```diff
@@ -823,6 +861,18 @@
...
+    /**
+     * Function used to search inside objects an object that contains pointer in bounding box or that contains pointerOnCanvas when painted
+     * @see {@link fabric.Canvas#_searchPossibleTargets}
+     * @param {Array} [objects] objects array to look into
+     * @param {Object} [pointer] x,y object of point coordinates we want to check.
+     * @return {fabric.Object} **top most object on screen** that contains pointer
+     */
+    searchPossibleTargets: function (objects, pointer) {
+      var target = this._searchPossibleTargets(objects, pointer);
+      return this.targets[0] || target;
```
Comment (ShaMan123) [inline src/canvas.class.js:873]: Here we decide what object gets selected, and this change is what makes all tests fail

**Comment by asturur** [comment]
@ShaMan123 are here all the changes to group? or is there something else i need to check?

**Comment by asturur** [comment]
What happened to the plan to do one change at time?
i see we have here both group changes  and nested selection changes all together.
I remember there was a unit of work, that was removing the need for changing object coordinates inside groups and wasn't this huge and seemed like easily reviewable.
Where is it now?

**Comment by asturur** [comment]
was it this one? 
https://github.com/fabricjs/fabric.js/pull/7599

**Comment by ShaMan123** [comment]
Here are all the changes
It is based on #7599 but I have worked on it since.
It is a major change.
I don't see a good way to break it into pieces.
You can checkout group.class.js and active selection.
But besides that... not a lot to do

**Comment by ShaMan123** [comment]
You can view it commit by commit...
The first 2 commits are #7599

**Comment by ShaMan123** [comment]
I didn't add nested selection logic
Just one line so I could test that everything works
here https://github.com/fabricjs/fabric.js/pull/7669#discussion_r807342442

**Comment by ShaMan123** [comment]
Well if it makes it easier I suggest not to review Group/Layer/ActiveSelection and review the rest.
It is much of what we already spoke of and that you reviewed in the google doc.

**Comment by asturur** [comment]
Let me read all, if i have a strong feeling for breaking it in pieces i ll tell you.
I know is a big change

**Comment by asturur** [comment]
There are improvements to the test suite that can be merged alone quickly, even if not used with the current test suite.

**Comment by ShaMan123** [comment]
I now understand why I experienced a perf hit.
`objectCaching` and `isOnACache` are suitablr rfor old group where it was wasteful to create caches for nested objects.
But this is not the case any longer. Nested selection triggers rendering of the selected object and it's group so we must address this.
In general I think of handling it somehow with logic including `subTargetCheck` but I am not sure it will be enough. Because there's a need for fine grained control over `objectCaching`.
What do you think?
e9549f5 is a perf commit (I thought the trouble came for there), it's important as well to reduce computation. But it means we need to expose a method on canvas to invalidate the chosen objects memo.

**Comment by ShaMan123** [comment]
> There are improvements to the test suite that can be merged alone quickly, even if not used with the current test suite.

Easy enough to cherry pick this out, should I?

**Comment by asturur** [comment]
> I now understand why I experienced a perf hit. `objectCaching` and `isOnACache` are suitable for old group. It was wasteful to create caches for nested objects. But this is not the case any longer. Nested selection triggers rendering of the selected object and it's group so we must address this. In general I think of handling it somehow with logic including `subTargetCheck` and `selectable|evented` but I am not sure it will be enough. Because there's a need for fine grained control over `objectCaching` in these cases. What do you think? [e9549f5](https://github.com/fabricjs/fabric.js/commit/e9549f515b1607ab56dd8734230dba77c59f599f) is a perf commit (I thought the trouble came for there), it's important as well to reduce computation. But it means we need to expose a method on canvas to invalidate the chosen objects memo.
> 
> We can turn `subTargetCheck` to false by default and then once the dev turns it on they will have a chance to read about this and handle `objectCaching` themseleves

So i have just briefly read the code changes, and is early for me to comment on.
Today i m sure i find time to review the canvas class changes and reason on them, those seems small, self contained, and harmless.

Caching logic may need to change entirely, yes. That is why i m mostly sure we will need to spend energy in breaking this PR in pieces that can be eventually reverted singularly.

I'm not taking any other tasks before this is fleshed out, so that we can do it.

**Comment by ShaMan123** [comment]
Alright, it might be that `objectCaching` need to become `none|always|optimized`

**Comment by ShaMan123** [comment]
reverted test suite changes 6b53d3c and checked out into #7691

File: src/shapes/object.class.js [hunk 5]
```diff
@@ -974,9 +974,7 @@
      * @returns {number}
      */
     getTotalAngle: function () {
-      return this.group ?
-        fabric.util.qrDecompose(this.calcTransformMatrix()).angle :
-        this.angle;
+      return fabric.util.qrDecompose(this.calcTransformMatrix()).angle;
```
Comment (asturur) [inline src/shapes/object.class.js:977]: is there any reason why we want this to not use shortcuts?

Comment (ShaMan123) [reply src/shapes/object.class.js:977]: this was something I saw when I merged  #7726 so I added it to there....
Shortcuts should be used

**Comment by ShaMan123** [comment]
@asturur this is stable

**Comment by ShaMan123** [comment]
### Old Group origin

image
image
image

**Comment by ShaMan123** [comment]
### New Group origin
This is why the group tests are failing. They should they were completely WRONG

image
image

image

**Comment by ShaMan123** [comment]
Now the question is if `ActiveSelection` should have originX/Y set to center by default or not.

**Comment by asturur** [comment]
image

I wish group were containers with no render properties.

**Comment by ShaMan123** [comment]
> image
> 
> I wish group were containers with no render properties.

Why?

**Comment by ShaMan123** [comment]
@asturur if you want to break off another simple chuck look at the change I made to collection

**Comment by ShaMan123** [comment]
So I will stop committing more stuff here so you can work.
To say the truth I think this PR is ready enough to be merged except for more tests on group for origin, layout, visual tests

**Comment by asturur** [comment]
> > image
> > I wish group were containers with no render properties.
> 
> Why?

To keep thing simpler. A container does the container and try to keep the scope limited to that

**Comment by ShaMan123** [comment]
We're talking about this
https://github.com/fabricjs/fabric.js/blob/9387fe65da82d26d7414e5c5fa0c81c2016157cd/src/shapes/group.class.js#L397
and svg export which is a small function
https://github.com/fabricjs/fabric.js/blob/9387fe65da82d26d7414e5c5fa0c81c2016157cd/src/shapes/group.class.js#L739

And perhaps svg import

That's it.

I see it as crucial, defaults remain as old group.

**Comment by ShaMan123** [comment]
There are 3 broken tests that are related to `searchPossibleTargets`
To see them do the following
return `target` from `searchPossibleTargets`
mark `Group.subTargetCheck` false as default and perhaps `ActiveSelection` as well
This reverts fabric to old behavior.
You'll see 2 or more tests failing only because of `subTargetCheck` default change.

<details>
<summary>failing tests</summary>
not ok 129 fabric.Canvas > findTarget with subTargetCheck
  ---
  message: should return the rect
  severity: failed
  actual  : undefined
  expected: rect#176
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:804:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
  ---
  message: should return the rect2
  severity: failed
  actual  : undefined
  expected: rect#177
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:814:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
ok 130 fabric.Canvas > findTarget with subTargetCheck and canvas zoom
not ok 131 fabric.Canvas > findTarget with subTargetCheck on activeObject
  ---
  message: should return the rect
  severity: failed
  actual  : undefined
  expected: rect#185
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:909:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
  ---
  message: multiple calls to subtarget should not add more to targets
  severity: failed
  actual  : 0
  expected: 1
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:923:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
not ok 132 fabric.Canvas > findTarget with subTargetCheck on activeObject and preserveObjectStacking true
  ---
  message: should return the rect
  severity: failed
  actual  : undefined
  expected: rect#188
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:940:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
  ---
  message: multiple calls to subtarget should not add more to targets
  severity: failed
  actual  : 0
  expected: 1
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:954:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
ok 133 fabric.Canvas > findTarget with perPixelTargetFind
ok 134 fabric.Canvas > findTarget with perPixelTargetFind in nested group
not ok 135 fabric.Canvas > findTarget on activegroup
  ---
  message: Should return the rect inside activegroup
  severity: failed
  actual  : undefined
  expected: rect#201
  stack: |
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\canvas.js:1084:12)
        at runTest (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2411:35)
        at Test.run (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2394:9)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2645:16
        at processTaskQueue (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1966:26)
        at C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:1970:13
  ...
</details>

**Comment by ShaMan123** [comment]
I got it 
tests 131, 132, 135 fail because Group doesn't render selected objects and let's canvas do it, see `_renderObjects`
test 129 has been fixed.
All tests pass after the revert suggested

**Comment by ShaMan123** [comment]
I thought I solved positioning with originX/Y but I didn't.
I think I pushed a regression somehow.
It is very confusing for so I will need you help with it.

**Comment by asturur** [comment]
The changes to controls12 would be mostly fine if it wasn't that the green controls have a weird angle that isn't part of any of the 2 possible bounding boxes. wrong per wrong isn't a big of an issues, but pointing out just in case you did not notice it.

I was working actively on group.class and collection.mixin, i noticed you push a bunch in the last hour.
Are things stable now or are you still refactoring some code?

File: src/shapes/group.class.js [hunk 6]
```diff
@@ -396,10 +396,15 @@
...
         //  render fill/stroke courtesy of rect
-        //fabric.Rect.prototype._render.call(this, ctx);
+        this._hasFillStroke() && fabric.Rect.prototype._render.call(this, ctx);
         this._renderObjects(ctx);
       },
 
+      _hasFillStroke: function () {
+        return (this.fill && this.fill !== 'none' && this.fill !== 'transparent') ||
+          (this.stroke && this.stroke !== 'none' && this.stroke !== 'transparent' && this.strokeWidth);
+      },
```
Comment (asturur) [inline src/shapes/group.class.js:406]: those exists has this.hasFill and this.hasStroke.
The 'none' value i m not sure if is reported somewhere, but shouldn't exist. I know it can come from SVG but we usually fiter it out

**Comment by ShaMan123** [comment]
Code is stable.
Working on passing the tests

**Comment by ShaMan123** [comment]
@asturur when group is used as a clip path should it render it's children using the clip path rendering functions or not?
Take a look at clippath-9.
clippath-9

**Comment by ShaMan123** [comment]
> The changes to controls12 would be mostly fine if it wasn't that the green controls have a weird angle that isn't part of any of the 2 possible bounding boxes. wrong per wrong isn't a big of an issues, but pointing out just in case you did not notice it.
> 
> I was working actively on group.class and collection.mixin, i noticed you push a bunch in the last hour. Are things stable now or are you still refactoring some code?

controls12 looks same to me besides the color of the bounding rect of the groups.
It is wrong, I think it belongs to the oCoords work. I'm staying off.

**Comment by asturur** [comment]
yes don't worry that is not working anyway, do not waste energy.

For the clipPath.
yes the shape of the group is supposed to define the clipPath.
What do you see?

**Comment by ShaMan123** [comment]
look at the difference in quality, it's blurry and has some lines going through it

**Comment by asturur** [comment]
> look at the difference in quality, it's blurry and has some lines going through it

yes blurry yes.
i think probably something from clippath should go away to simplify use cases and fix blurryness.
I have a PR open for it, i hit some walls but i did not dig deeper.
Caching the clipPath is probably an error, and that is the first thing i removed entirely, i m sure that with the correct transforms i can fix the issues i encountered

**Comment by ShaMan123** [comment]
I think this one is my group's fault

**Comment by asturur** [comment]
i remember it was always there

File: src/shapes/object.class.js [hunk 7]
```diff
@@ -1113,7 +1113,7 @@
      * @returns Boolean
      */
     hasStroke: function() {
-      return this.stroke && this.stroke !== 'transparent' && this.strokeWidth !== 0;
+      return this.stroke && this.stroke !== 'transparent' && this.stroke !== 'none' && this.strokeWidth !== 0;
```
Comment (asturur) [inline src/shapes/object.class.js]: none is not an accepted color and shouldn't be. none renders black has any string that is not a color and can't be parsed

File: src/shapes/active_selection.class.js [hunk 8]
```diff
@@ -24,58 +24,37 @@
...
-      });
-      newGroup._objects = objects;
-      if (!this.canvas) {
-        return newGroup;
-      }
-      var canvas = this.canvas;
-      canvas.add(newGroup);
-      canvas._activeObject = newGroup;
-      newGroup.setCoords();
-      return newGroup;
```
Comment (asturur) [inline src/shapes/active_selection.class.js:78]: why do we want to get rid of `toGroup` method on activeSelection?

File: src/shapes/object.class.js (hunk 7 above)
Comment (ShaMan123) [reply src/shapes/object.class.js]: I removed it

File: src/shapes/active_selection.class.js (hunk 8 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:78]: Too hacky IMO
I aim to make it as simple as `new fabric.Group(canvas.getActiveObjects())`

**Comment by asturur** [comment]
Did you do any use of the insertAt non-splicing case?
i think since an old bugifx to support subclassing use cases, the non splicing is completely unused

**Comment by ShaMan123** [comment]
I have exposed it on group and updated it's tests.
But it can go.

**Comment by ShaMan123** [comment]
`groupSVGElements` is a pain 
I can't get it to work exactly as old group did, layout is different, need help

**Comment by ShaMan123** [comment]
> > look at the difference in quality, it's blurry and has some lines going through it
> 
> yes blurry yes. i think probably something from clippath should go away to simplify use cases and fix blurryness. I have a PR open for it, i hit some walls but i did not dig deeper. Caching the clipPath is probably an error, and that is the first thing i removed entirely, i m sure that with the correct transforms i can fix the issues i encountered

I think this is because of object caching inside group/group as clip path

**Comment by asturur** [comment]
@ShaMan123 if groupSvgElement does not work for a week or a month i don't care. is a detail. We will make it work.

As of now i m focused in extracting a part of this work in a mergeable unit.
I left the interactivity part out, and i m trying to keep things working mostly as before but introducing all the collection and group changes. Even just adding those pieces, is a large change.

I would like to write down instructions on how a group is used
Trying to get UTs to pass on the small subset of change, and figuring out why object caching is changed for group.

File: src/shapes/active_selection.class.js (hunk 8 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:78]: > Too hacky IMO I aim to make it as simple as `new fabric.Group(canvas.getActiveObjects())`

Working out of the box
https://codesandbox.io/s/jumshw

File: src/shapes/layer.class.js [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    __objectMonitor: function () {
+      //  noop
+    },
+
+    /**
+     * @private
+     */
+    __canvasMonitor: function () {
+      this._applyLayoutStrategy({ type: 'canvas_resize' });
```
Comment (ShaMan123) [inline src/shapes/layer.class.js:128]: I think we might want to trigger a layout from `Canvas.setDimensions` as well

File: src/shapes/active_selection.class.js (hunk 8 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:78]: http://jsfiddle.net/Lxjegow3/8/

**Comment by ShaMan123** [comment]
I have dropped old group's caching checks so the issue with caching probably lies there.
I didn't build new logic because as we discussed there's a need to rewrite it

File: src/shapes/group.class.js [hunk 10]
```diff
@@ -15,560 +21,872 @@
...
+        fabric.Collection.insertAt.call(this, objects, index, nonSplicing,
+          relativeToGroup ? this._onRelativeObjectAdded : this._onObjectAdded);
+        this._onAfterObjectsChange('added', Array.isArray(objects) ? objects : [objects]);
+      },
+
+      /**
+       * Remove objects
+       * @param {...fabric.Object} objects
+       */
+      remove: function () {
```
Comment (ShaMan123) [inline src/shapes/group.class.js:180]: add `removeRelative`

File: src/shapes/group.class.js [hunk 11]
```diff
@@ -15,560 +21,872 @@
...
+       * @private
+       * @param {fabric.Object} object
+       * @param {boolean} [removeParentTransform] true if object is in canvas coordinate plane
+       */
+      enterGroup: function (object, removeParentTransform) {
+        if (object.group) {
+          if (object.group === this) {
+            throw new Error('fabric.Group: duplicate objects are not supported inside group');
+          }
+          object.group.remove(object);
```
Comment (ShaMan123) [inline src/shapes/group.class.js:267]: Here I think it makes sense to preserve the old group's transform if `removeParentTransform` is true

**Comment by ShaMan123** [comment]
Alright,
Future thoughts regarding new selection logic:
If an object is selected logic target check should bubble up the tree to parents, siblings and relations in some manner.
Only after bubbling has reached it's course and is now at the top most ancestor of the active object should it iterate down the tree on all remaining objects.

Thoughts? @asturur @melchiar

File: src/shapes/group.class.js [hunk 12]
```diff
@@ -15,560 +21,872 @@
...
+      _set: function (key, value) {
+        var prev = this[key];
+        this.callSuper('_set', key, value);
+        if (key === 'canvas' && prev !== value) {
+          this.forEachObject(function (object) {
+            object._set(key, value);
+          });
+        }
+        if (key === 'layout' && prev !== value) {
+          this._applyLayoutStrategy({ type: 'layout_change', layout: value, prevLayout: prev });
```
[the remark this answers is not in the document]
Comment (ShaMan123) [reply src/shapes/group.class.js:137]: fixed by #7761

**Comment by ShaMan123** [comment]
@asturur I now understand your incline not to have fill/stroke for group. I was not aware of `backgroundColor` which is good enough for most use cases. So I don't oppose removing

**Comment by ShaMan123** [comment]
### Take a look at **THIS**!

The blue circles are rendered by a textbox and because of the rewrite with a few lines of code I got this working
Pay no notice to caching of course

ezgif com-gif-maker (13)

**Comment by ShaMan123** [comment]
apart from `_renderObjects` that has been completely consumed
